### ISSUE TITLE
How to solve warp cross-lane functions work in branching code

### ISSUE THREAD

**Issue by fileaccent** [ROCm-Developer-Tools/HIP#3172]
This is a sample example:
```
#include <iostream>
#include <hip/hip_runtime.h>
using namespace std;

__global__ void kernel() {
     int reg = threadIdx.x;
     if (reg % 2 == 1)
        printf("now: %d, reg_4: %d\n", reg, __shfl(reg, 4, 64));
     else
        printf("now: %d, reg_2: %d\n", reg, __shfl(reg, 2, 64));

}

int main () {
        kernel<<<1, 64>>>();
        hipDeviceSynchronize();
        return 0;
}
```

This is the answer
```
now: 0, reg_2: 2
now: 2, reg_2: 2
now: 4, reg_2: 2
now: 6, reg_2: 2
now: 8, reg_2: 2
now: 10, reg_2: 2
now: 12, reg_2: 2
now: 14, reg_2: 2
now: 16, reg_2: 2
now: 18, reg_2: 2
now: 20, reg_2: 2
now: 22, reg_2: 2
now: 24, reg_2: 2
now: 26, reg_2: 2
now: 28, reg_2: 2
now: 30, reg_2: 2
now: 32, reg_2: 2
now: 34, reg_2: 2
now: 36, reg_2: 2
now: 38, reg_2: 2
now: 40, reg_2: 2
now: 42, reg_2: 2
now: 44, reg_2: 2
now: 46, reg_2: 2
now: 48, reg_2: 2
now: 50, reg_2: 2
now: 52, reg_2: 2
now: 54, reg_2: 2
now: 56, reg_2: 2
now: 58, reg_2: 2
now: 60, reg_2: 2
now: 62, reg_2: 2
now: 1, reg_4: 0
now: 3, reg_4: 0
now: 5, reg_4: 0
now: 7, reg_4: 0
now: 9, reg_4: 0
now: 11, reg_4: 0
now: 13, reg_4: 0
now: 15, reg_4: 0
now: 17, reg_4: 0
now: 19, reg_4: 0
now: 21, reg_4: 0
now: 23, reg_4: 0
now: 25, reg_4: 0
now: 27, reg_4: 0
now: 29, reg_4: 0
now: 31, reg_4: 0
now: 33, reg_4: 0
now: 35, reg_4: 0
now: 37, reg_4: 0
now: 39, reg_4: 0
now: 41, reg_4: 0
now: 43, reg_4: 0
now: 45, reg_4: 0
now: 47, reg_4: 0
now: 49, reg_4: 0
now: 51, reg_4: 0
now: 53, reg_4: 0
now: 55, reg_4: 0
now: 57, reg_4: 0
now: 59, reg_4: 0
now: 61, reg_4: 0
now: 63, reg_4: 0
```

The odd term should print 4. 
How do I get the right answer?

The code runs on Vega 10 XTX [Radeon Vega Frontier Edition], rocm version 5.3.0

**Comment by yxsamliu** [comment]
This is due to an LLVM pass SimplifyCFG. Before the pass:

`define protected amdgpu_kernel void @_Z6kernelPi(ptr addrspace(1) nocapture noundef writeonly %y.coerce) local_unnamed_addr #1 {
entry:
  %0 = tail call i32 @llvm.amdgcn.workitem.id.x(), !range !6
  %rem = and i32 %0, 1
  %cmp.not = icmp eq i32 %rem, 0
  br i1 %cmp.not, label %if.else, label %if.then

if.then:                                          ; preds = %entry
  %1 = tail call i32 @llvm.amdgcn.mbcnt.lo(i32 -1, i32 0)
  %2 = tail call i32 @llvm.amdgcn.mbcnt.hi(i32 -1, i32 %1)
  %and.i = shl i32 %2, 2
  %add.i = and i32 %and.i, -256
  %shl.i = or i32 %add.i, 16
  %3 = tail call i32 @llvm.amdgcn.ds.bpermute(i32 %shl.i, i32 %0)
  %idxprom = zext i32 %0 to i64
  %arrayidx = getelementptr inbounds i32, ptr addrspace(1) %y.coerce, i64 %idxprom
  store i32 %3, ptr addrspace(1) %arrayidx, align 4, !tbaa !7
  br label %if.end

if.else:                                          ; preds = %entry
  %4 = tail call i32 @llvm.amdgcn.mbcnt.lo(i32 -1, i32 0)
  %5 = tail call i32 @llvm.amdgcn.mbcnt.hi(i32 -1, i32 %4)
  %and.i11 = shl i32 %5, 2
  %add.i12 = and i32 %and.i11, -256
  %shl.i13 = or i32 %add.i12, 8
  %6 = tail call i32 @llvm.amdgcn.ds.bpermute(i32 %shl.i13, i32 %0)
  %idxprom4 = zext i32 %0 to i64
  %arrayidx5 = getelementptr inbounds i32, ptr addrspace(1) %y.coerce, i64 %idxprom4
  store i32 %6, ptr addrspace(1) %arrayidx5, align 4, !tbaa !7
  br label %if.end

if.end:                                           ; preds = %if.else, %if.then
  ret void
`
After the pass:
`define protected amdgpu_kernel void @_Z6kernelPi(ptr addrspace(1) nocapture noundef writeonly %y.coerce) local_unnamed_addr #1 {
entry:
  %0 = tail call i32 @llvm.amdgcn.workitem.id.x(), !range !6
  %rem = and i32 %0, 1
  %cmp.not = icmp eq i32 %rem, 0
  %1 = tail call i32 @llvm.amdgcn.mbcnt.lo(i32 -1, i32 0)
  %2 = tail call i32 @llvm.amdgcn.mbcnt.hi(i32 -1, i32 %1)
  %and.i11 = shl i32 %2, 2
  %add.i12 = and i32 %and.i11, -256
  %idxprom4 = zext i32 %0 to i64
  %arrayidx5 = getelementptr inbounds i32, ptr addrspace(1) %y.coerce, i64 %idxprom4
  %. = select i1 %cmp.not, i32 8, i32 16
  %shl.i13 = or i32 %add.i12, %.
  %3 = tail call i32 @llvm.amdgcn.ds.bpermute(i32 %shl.i13, i32 %0)
  store i32 %3, ptr addrspace(1) %arrayidx5, align 4, !tbaa !7
  ret void

`
SimplifyCFG sinks @llvm.amdgcn.mbcnt.lo and @llvm.amdgcn.mbcnt.hi out of the branches since it looks like they do not depend on the CFG. This is wrong since mbcnt counts the bits of active lanes, which depends on CFG.

This is a known issue https://discourse.llvm.org/t/rfc-introduce-cross-lane-function-attribute-to-prevent-merging-call-of-cross-lane-functions/62148/2

Basically, the LLVM community is trying to address this issue. However, it will take a long time since it involves some fundamental changes.

Currently, this is treated as a limitation of HIP, that is: cross-lane functions e.g. shfl_* do not work in divergent control flow. The workaround is to avoid using cross-lane functions in divergent control flow, e.g. the above sample can be workaround by

`
__global__ void kernel() {
     int reg = threadIdx.x;
     int tmp;
     if (reg % 2 == 1)
        tmp=4;
     else
        tmp=2;
     printf("now: %d, reg_4: %d\n", reg, __shfl(reg, tmp, 64));

}

`

**Comment by yxsamliu** [comment]
I am trying to fix this issue by https://reviews.llvm.org/D144756

A workaround is to add -mllvm -hoist-common-insts=0 -mllvm -sink-common-insts=0 to your compilation command.

**Comment by yxsamliu** [comment]
This issue need another patch https://reviews.llvm.org/D145072

**Comment by jayfoad** [comment]
> This is wrong since mbcnt counts the bits of active lanes, which depends on CFG.

No, the result of mbcnt does not depend on which lanes are active.

**Comment by ssahasra** [comment]
```
     if (reg % 2 == 1)
        printf("now: %d, reg_4: %d\n", reg, __shfl(reg, 4, 64));
     else
        printf("now: %d, reg_2: %d\n", reg, __shfl(reg, 2, 64));
```
The way I see it, the output is undefined for odd-numbered threads. Doesn't __shfl() require that accessed threads are active? In that case, if only odd-numbered threads are active, and they try to read from lane ID 4, they are reading from an inactive thread.

Changing "4" to any odd number produces the correct result, but changing it to any even number produces zero.
